developer_manual/digging_deeper/groupware/calendar_provider.rst

```diff
@@ -0,0 +1,637 @@
+
+.. code-block:: php
+
+    <?php
+    
+    function getLastModified() {
+        return time();
+    }
+
+The last time the calender is modified allows clients to optimize their requests. This method should return the corresponding unix timestamp.
+
+A fallback is to provide the value ``null`` as return value. This tells that the last modification time is not known at the moment.
+
+Entries in the calendar -- ICollection
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The interface ``\Sabre\DAV\ICollection`` defines methods to access children of the current node. For calendars, the children are in fact the appointments stored within the calendar. Again, some methods are already covered, so here only the required methods are implemented.
+
+All calendar entries do have a unique name. This is just a plain string. Typically these are named as ``.ics`` files.
+
+.. code-block:: php
+
+    <?php
+    
+    function createFile($name, $data = null) {
+        return null;
+        // return "\"$etag\"";
+    }
+
+This method is used to store new appointments to the calendar. One could return return an ETag of the calendar appointment as a string that contains double quotes as sketched in the comment.
+
+.. code-block:: php
+
+    <?php
+    
+    function childExists($name) {
+        // Check if the value of $name represents a valid calendar entry name.
+        return $name === 'test.ics';
+    }
+
+The ``childExists`` method checks if a certain element is present in the calendar. In this example we have only one event called ``test.ics``.
+
+.. code-block:: php
+
+    <?php
+    
+    function getChild($name) {
+        if ($this->childExists($name)) {
+            return new CalendarObject($this, $name);
+        }
+    }
+
+This will pack an calendar entry into its own object as described earlier.
+
+The method allows to request a specific entry and extract it from the calendar.
+
+.. code-block:: php
+
+    <?php
+    
+    function getChildren() {
+        // Get the list of calendar entries
+        $children = ['test.ics'];
+
+        // Obtain the calendar objects for each of them
+        $children = array_map(function ($childName) using ($this) { return $this->getChild($childName); });
+        
+        return $children;
+    }
+
+Finally, there is a class to fetch all appointments of a calendar.
+
+.. note:: For the sake of simplicity, here only a static array is used. One could however query a database or the file system for a variable number of entries in the calendar.
+
+Querying the calendar -- ICalendarObjectContainer
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For calendar entries it makes little sense to query all and sort on the client the relevant ones. Instead, the client requests a certain set of objects (like the last 90 days) and the server will do the filtering. This can be achieved by the ``\Sabre\CalDAV\ICalendarObjectContainer`` interface.
```

Comment:
```suggestion
It would be very resource intensive to request all events of a calendar only to then discard most of them during filtering. Instead, the client requests a certain set of objects (like the last 90 days) and the server will do the filtering. This can be achieved by the ``\Sabre\CalDAV\ICalendarObjectContainer`` interface.
```